### PR TITLE
[RLlib] IMPALA trainer tf

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: Avnish <avnishnarayan@gmail.com>

## Why are these changes needed?

Adds the impala tf trainer and enables usage of the learner group and RLModule apis in IMPALA.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/testing/utils.py
Comment (avnishn) [inline rllib/core/testing/utils.py:164]: lets just go ahead and not merge this. Its not a deep enough abstraction.

File: rllib/core/learner/tf/tf_learner.py
Comment (avnishn) [inline rllib/core/learner/tf/tf_learner.py:158]: Im not sure if this is doing what I expect it to do .... Will need reworking.

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (avnishn) [inline rllib/algorithms/impala/tf/impala_tf_learner.py:150]: we need this for keeping parity between the old and new IMPALA impls

File: rllib/algorithms/algorithm_config.py
Comment (avnishn) [inline rllib/algorithms/algorithm_config.py:2759]: passing grad clip to the learner

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [inline rllib/algorithms/impala/impala.py:153]: need to provide the rollout frag length in order to do vtrace. But at the same time what if someone does complete episodes mode?

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [inline rllib/algorithms/impala/impala.py:732]: not inheriting from base class's function, so we need to add a pg bundle for the learner group

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [inline rllib/algorithms/impala/impala.py:787]: need a way to sample even if sampling from local worker.

File: rllib/algorithms/impala/tests/test_impala_off_policyness.py
Comment (avnishn) [inline rllib/algorithms/impala/tests/test_impala_off_policyness.py:45]: eager tracing is disabled when using rlm api for now

File: rllib/algorithms/impala/tests/tf/test_impala_learner.py
Comment (avnishn) [inline rllib/algorithms/impala/tests/tf/test_impala_learner.py:38]: pretty much copied this test from the ppo branch

File: rllib/algorithms/impala/tests/tf/test_vtrace_v2.py
Comment (avnishn) [inline rllib/algorithms/impala/tests/tf/test_vtrace_v2.py:21]: this is the same test that we merged and so im forgetting if there is a reason that it looks like we added this as a new test.

File: rllib/core/learner/tf/tf_learner.py
Comment (avnishn) [inline rllib/core/learner/tf/tf_learner.py:98]: new addition of gradient clipping logic

**Review by avnishn — COMMENTED** [review]
leaving some self review comments

**Review by avnishn — COMMENTED** [review]
Separate logic for learner and non learner into different functions in training step

File: rllib/algorithms/impala/impala.py
Comment (kouroshHakha) [reply rllib/algorithms/impala/impala.py:807]: only if they are relevant in the new stack.

File: rllib/algorithms/impala/impala.py
Comment (kouroshHakha) [reply rllib/algorithms/impala/impala.py:744]: yes this is a tune / ray core feature request.

**Review by kouroshHakha — COMMENTED** [review]
ok releasing my comments now. Thanks.

File: rllib/algorithms/impala/tests/test_impala_off_policyness.py
Comment (kouroshHakha) [reply rllib/algorithms/impala/tests/test_impala_off_policyness.py:45]: This is more readable than `n_aggr_worker = perm[0]` and `enable_rlm = perm[1]`

**Comment by kouroshHakha** [comment]
also lint is failing.

**Comment by avnishn** [comment]
lint is failing because i need to add a main fn to one of my tests

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py:153]: thats fair. I thought that would be a solution as well but that it would be invalid because we don't like modifying our configs inside of the validate fn. Then again, we aren't changing the hparams at all, so I guess there shouldn't be a problem.

Comment (kouroshHakha) [reply rllib/algorithms/impala/impala.py:153]: you should raise an error if it's already set or something and then set it. This way it's explicit and there is not magic behind the scene.

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py:787]: It was, its just there was never a reason to run impala with 0 workers before this, and definitely would hang if there were 0 workers.

File: rllib/algorithms/impala/tests/tf/test_vtrace_v2.py
Comment (avnishn) [reply rllib/algorithms/impala/tests/tf/test_vtrace_v2.py:21]: I did some digging ... I renamed the test and moved it to somewhere else in the file tree.

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py:744]: gotcha .. I'll add a todo here

File: rllib/algorithms/algorithm_config.py
Comment (avnishn) [reply rllib/algorithms/algorithm_config.py:2759]: fixed

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:118]: Over here I'm borrowing from our original implementation. Why don't we try it out during benchmarking and see if there is a difference.

Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:118]: Update: I tried it out and the learning collapsed. I couldn't find the set of optimal hparams to make it work when using mean vs sum. 

I think at this point its not a useful change, especially since we already have benchmarked this algorithm with this loss on many envs, and making this change would require rebenchmarking to find the optimal hparams.

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:46]: does learner_hps have a validate function? I don't think it does. I'm going to leave this as is unless we have one.

File: rllib/core/learner/tf/tf_learner.py
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:101]: This doesn't seem to exist for as an option for all algorithms just yet. Otherwise it would probably be apart of algorithm_config options.

Going to leave as is since I moved the grad_clip configuration into impala.py from algorithm_config.py

File: rllib/core/learner/tf/tf_learner.py
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:106]: kk

File: rllib/core/learner/tf/tf_learner.py
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:155]: its not necessary I was mistaken :)

File: rllib/tuned_examples/impala/cartpole-impala.yaml
Comment (avnishn) [reply rllib/tuned_examples/impala/cartpole-impala.yaml:12]: its probably fine for now. As we add more algorithms we can add this as well.

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py]: ah they are not relevant in the new stack.

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (kouroshHakha) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:72]: Don't these values get materialized at algorithm_config validation? Shouldn't we check their validity there?

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (kouroshHakha) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:142]: I mean you could extract batch_size (B) and then do only one tf.reduce_sum(xxxx) and then do loss = sum / B in order to not do the same computation twice.

File: rllib/algorithms/impala/tests/test_impala_off_policyness.py
Comment (kouroshHakha) [inline rllib/algorithms/impala/tests/test_impala_off_policyness.py:45]: We can probably enable eager tracing now.

**Review by kouroshHakha — APPROVED** [review]
looks good. Some final comments.

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py:1037]: actually workers_that_need_updates is always going be = {0} since sampling will always be blocking in this case.

I can just lop off the first part of the statement.

Comment (avnishn) [reply rllib/algorithms/impala/impala.py:1037]: done

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py:404]: done

File: rllib/algorithms/impala/impala.py
Comment (avnishn) [reply rllib/algorithms/impala/impala.py:826]: I just copied this verbatim. Its a good explanation.

Comment (avnishn) [reply rllib/algorithms/impala/impala.py:826]: done

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:72]: resolved offline, adding a todo for validate function in learner_hps.

File: rllib/algorithms/impala/tf/impala_tf_learner.py
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:142]: fixed

File: rllib/core/learner/tf/tf_learner.py
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:103]: added a todo

File: rllib/tuned_examples/impala/cartpole-impala.yaml
Comment (avnishn) [reply rllib/tuned_examples/impala/cartpole-impala.yaml]: I think that we can do this after we have merged the model catalog stack. And then also we can shift the release tests to the new api.

File: rllib/algorithms/impala/tests/test_impala_off_policyness.py
Comment (avnishn) [reply rllib/algorithms/impala/tests/test_impala_off_policyness.py:45]: re-enabled

**Comment by avnishn** [comment]
okie all the comments are resolved, lets just wait for tests to pass and merge.

**Pull request merged**
